packages/api/db/migration/20240207214919_create_animal_union_batch_view_modify_animal_batch_tables.js

```diff
@@ -0,0 +1,106 @@
+/*
+ *  Copyright 2024 LiteFarm.org
+ *  This file is part of LiteFarm.
+ *
+ *  LiteFarm is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  LiteFarm is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
+ */
+
+/**
+ * @param { import("knex").Knex } knex
+ * @returns { Promise<void> }
+ */
+export const up = async function (knex) {
+  try {
+    await knex.schema.alterTable('animal', (table) => {
+      table.string('photo_url');
+      table.dropChecks(['name_identifier_check']);
+    });
+
+    await knex.schema.alterTable('animal_batch', (table) => {
+      table.string('photo_url');
+      table.string('name').nullable().alter();
+    });
+
+    // Create function to assign internal_identifier to a record (NEW) to be inserted
+    await knex.raw(`
+      CREATE VIEW animal_union_batch AS
+      SELECT
+        *,
+        ROW_NUMBER() OVER (PARTITION BY farm_id ORDER BY created_at)::INTEGER AS internal_identifier
+      FROM (
+        SELECT
+          id,
+          farm_id,
+          default_type_id,
+          custom_type_id,
+          default_breed_id,
+          custom_breed_id,
+          name,
+          identifier,
+          (SELECT array_agg(animal_group_id) FROM animal_group_relationship WHERE animal_id = a.id) AS related_group_ids,
+          1 AS count,
+          FALSE AS batch,
+          deleted,
+          created_at
+        FROM
+          animal a
+
+        UNION ALL
+
+        SELECT
+          id,
+          farm_id,
+          default_type_id,
+          custom_type_id,
+          default_breed_id,
+          custom_breed_id,
+          name,
+          NULL AS identifier,
+          (SELECT array_agg(animal_group_id) FROM animal_batch_group_relationship WHERE animal_batch_id = ab.id) AS related_group_ids,
+          count,
+          TRUE AS batch,
+          deleted,
+          created_at
+        FROM
+          animal_batch ab
+      ) animal_union_batch
+      ORDER BY
+        created_at;
+    `);
+  } catch (error) {
+    console.error('Error in migration up:', error);
+    throw error; // Rethrow the error to ensure the migration fails
+  }
+};
+
+/**
+ * @param { import("knex").Knex } knex
+ * @returns { Promise<void> }
+ */
+export const down = async (knex) => {
+  await knex.schema.dropView('animal_union_batch');
+
+  await knex.schema.alterTable('animal', (table) => {
+    table.dropColumn('photo_url');
+  });
+
+  // table.check does not work in alterTable
```

Comment:
Do you know why this is by chance? You can get `table.check()` to work -- but only with an empty binding array (something I saw suggested on an open issue on the knex repo):

```js
table.check('name IS NOT NULL OR identifier IS NOT NULL', [], 'name_identifier_check');
 ```
 
This was driving me nuts just now on a migration I was trying to write. Do you know why the ?? interpolation works in `createTable` but not `alterTable`?

Reply:
Ah! So table.check still works!
I couldn't find any info other than that `alterTable` may have some limitations  😢